audb/core/stream.py

```diff
@@ -0,0 +1,701 @@
+            that is not part of the database
+        ValueError: if a non-supported ``bit_depth``,
+            ``format``,
+            or ``sampling_rate``
+            is requested
+        RuntimeError: if a flavor is requested,
+            but the database contains media files,
+            that don't contain audio,
+            e.g. text files
+
+    Examples:
+        >>> import numpy as np
+        >>> np.random.seed(1)
+        >>> db = audb.stream(
+        ...     "emodb",
+        ...     "files",
+        ...     version="1.4.1",
+        ...     batch_size=4,
+        ...     shuffle=True,
+        ...     only_metadata=True,
+        ...     full_path=False,
+        ...     verbose=False,
+        ... )
+        >>> next(db)
+                                         duration  speaker transcription
+        file
+        wav/14a05Fb.wav 0 days 00:00:03.128687500       14           a05
+        wav/15a05Eb.wav 0 days 00:00:03.993562500       15           a05
+        wav/12a05Nd.wav    0 days 00:00:03.185875       12           a05
+        wav/13a07Na.wav 0 days 00:00:01.911687500       13           a07
+
+    """
+    if version is None:
+        version = latest_version(name)
+
+    flavor = Flavor(
+        bit_depth=bit_depth,
+        channels=channels,
+        format=format,
+        mixdown=mixdown,
+        sampling_rate=sampling_rate,
+    )
+    db_root = database_cache_root(name, version, cache_root, flavor)
+
+    deps = dependencies(
+        name,
+        version=version,
+        cache_root=cache_root,
+        verbose=verbose,
+    )
+
+    if table not in deps.table_ids:
+        msg = error_message_missing_object("table", [table], name, version)
+        raise ValueError(msg)
+
+    with FolderLock(db_root):
+        # Start with database header without tables
+        db, backend_interface = load_header_to(
+            db_root,
+            name,
+            version,
+            flavor=flavor,
+            add_audb_meta=True,
+        )
+
+        # Misc tables required by schemes of requested table
+        misc_tables = _misc_tables_used_in_table(db[table])
+
+        # Load table files
+        _load_files(
+            misc_tables + [table],
+            "table",
+            backend_interface,
+            db_root,
+            db,
+            version,
+            None,
+            deps,
+            Flavor(),
+            cache_root,
+            False,  # pickle_tables
+            num_workers,
+            verbose,
+        )
+
+        # Load misc tables completely
+        for misc_table in misc_tables:
+            table_file = os.path.join(db_root, f"db.{misc_table}")
+            db[misc_table].load(table_file)
+
+    if os.path.exists(os.path.join(db_root, f"db.{table}.parquet")):
+        database_iterator_object = DatabaseIteratorParquet
+    else:
+        database_iterator_object = DatabaseIteratorCsv
+
+    database_iterator = database_iterator_object(
```

Comment:
Most of the kwargs are passed to the `DatabaseIterator*` classes unaltered. 
Would it make sense to take care of that and not having to copy them one by one?

I have posted a simple patch to illustrate what I mean below (apply with `patch audb/core/stream.py stream.py.patch ` if you want to try it out).

Probably there are better approaches.  There seem to be decoarators from `functools` that might do this in a nicer fashion but I do not  see how well that works, see [here](https://discuss.python.org/t/get-exact-function-call-keyword-arguments-in-order/38155/3). 

This would help to make the caling at the bottom of the function nicer - but you have to manipulate the kwargs dict.  So both have their pros and cons.


```diff 
diff --git a/audb/core/stream.py b/audb/core/stream.py
index 3bb2b8d..9067421 100644
--- a/audb/core/stream.py
+++ b/audb/core/stream.py
@@ -614,9 +614,14 @@ def stream(
         wav/13a07Na.wav 0 days 00:00:01.911687500       13           a07
 
     """
+
     if version is None:
         version = latest_version(name)
 
+    # used to pass on to streaming iterator constructor
+    # name and table are args that are not passed on
+    kwargs =  dict((k, v) for (k, v) in locals().items() if k not in ['name', 'table'])
+
     flavor = Flavor(
         bit_depth=bit_depth,
         channels=channels,
@@ -672,30 +677,9 @@ def stream(
             table_file = os.path.join(db_root, f"db.{misc_table}")
             db[misc_table].load(table_file)
 
+    # update version kwarg before passing on
+    kwargs['version'] = version
     if os.path.exists(os.path.join(db_root, f"db.{table}.parquet")):
-        database_iterator_object = DatabaseIteratorParquet
+        return DatabaseIteratorParquet(db, table, **kwargs)
     else:
-        database_iterator_object = DatabaseIteratorCsv
-
-    database_iterator = database_iterator_object(
-        db,
-        table,
-        version=version,
-        map=map,
-        batch_size=batch_size,
-        shuffle=shuffle,
-        buffer_size=buffer_size,
-        only_metadata=only_metadata,
-        bit_depth=bit_depth,
-        channels=channels,
-        format=format,
-        mixdown=mixdown,
-        sampling_rate=sampling_rate,
-        full_path=full_path,
-        cache_root=cache_root,
-        num_workers=num_workers,
-        timeout=timeout,
-        verbose=verbose,
-    )
-
-    return database_iterator
+        return DatabaseIteratorCsv(db, table, **kwargs)


```

Reply:
Good point, I'm also very annoyed by passing on all the arguments.

I implemented now a slightly modified version of your solution. `kwargs['version'] = version` is not needed as `version` is already updated before assigning `kwargs`.

I also updated `DatabaseIteratorCsv` and `DatabaseIteratorParquet` as I could remove their whole `__init__()` methods, which are also only passing on arguments.

Reply:
>> I implemented now a slightly modified version of your solution. kwargs['version'] = version is not needed as version is already updated before assigning kwargs.

I know, you can probably put the code to gather simply after version is updated if it is None.  Then it would have the right value from the start.  But let's first see how you implemented it.

>> I also updated DatabaseIteratorCsv and DatabaseIteratorParquet as I could remove their whole __init__() methods, which are also only passing on arguments.

Afaics the parametrization is moved to the superclass entirely. This keeps the csv/parquet specific amount of code nicely small. For me this is also greatly facilitates reading the code. 

I think this is nice and I will resolve this conversation.